keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1182 @@
+Kubernetes system to fix all the node drain issues we mentioned earlier. The name, NodeMaintenance
+is subject to change; please see the alternatives section for other suggestions.
+
+To support workload migration, a new controller should be introduced to observe the NodeMaintenance
+objects and then mark pods for migration or removal with a condition. The pods would be selected
+according to the node at first, but the selection mechanism can be extended later. Controllers can
+then implement the migration. The advantage of this approach is that controllers do not have to be
+aware of the NodeMaintenance object (no RBAC changes required). They only have to observe pods they
+own and react by migrating them. The first candidate is a deployment controller, since its workloads
+support surging to another nodes. This would help to eliminate downtime not only for single replica
+applications, but for HA applications as well. 
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I also want to observe the node drain via the API and check on
+its progress. I also want to be able to discover workloads that are blocking the node drain.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another.
+
+#### Story 3
+
+Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
+drain a node using the same API and provide a similar experience to the CLI counterpart.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Kubectl
+
+`kubectl drain` command will be changed to create a NodeMaintenance object after marking node
+unschedulable. We will also change the implementation to skip applications that support
+workload migration. At first, we will just check deployments with a positive
+`.spec.strategy.rollingUpdate.maxSurge` value. To make this extensible, we propose a new Surge API
+that could be checked before disrupting any workload (see below). If the cluster,
+doesn't support the NodeMaintenance API, kubectl will perform the node drain in a backwards
```

Comment:
Will `kubectl` fail if you set `--node-maintenance` and it doesn't actually support that? Seems like it should in that case so the operator knows.

Reply:
We need to make `kubectl drain` compatible with the old behaviour, so we cannot fail there in this scenario. I think I will remove the `--node-maintenance` option from `kubectl uncordon` for now, as it doesn't make much sense in case the selected nodes do not overlap with the `NodeMaintenance` object.

Reply:
In the current version, NodeMaintenance object and controller is taking a full responsibility for cordoning and uncordoning the nodes. The kubectl commands should not be necessary anymore. 

I like the idea of showing warnings when users use `kubectl cordon` and `kubectl uncordon`. These commands will become noop as the node maintenance controller will stomp over their changes.